Add template methods into client

dudoslav opened the pull request:
Client can now send/recv any data type

dudoslav committed:
Add template methods into client

Client can now send/recv any data type

tomsik68 commented on client.hpp:
```diff
@@ -13,11 +18,40 @@ class Client {
+
+  template<class T>
+  void write(const T& data) {
+    write(data.data(), data.size()); 
```
Converting size to length may not always be correct. 
Consider this:
```
std::vector<uint32_t> vec{1u};
client << vec;
```
calls `client.write(&vec[0], 1); `, which writes one byte instead of one vector element. `send`'s `length` parameter is in bytes, so `data.size()` should be multiplied by size of one element of T(assuming T is a container).

tomsik68 committed:
add test for vector write

Sending vector<T> where sizeof(T) != 1 causes a major issue.
As `send` takes length in bytes and T.size() returns number of elements
in vector, vectors where   vector.size() != sizeof(T) * vector.size()
are transferred incorrectly

Signed-off-by: Tomas Jasek <tomsik68@gmail.com>

tomsik68 commented on client.hpp:
```diff
@@ -13,11 +18,40 @@ class Client {
+  template<typename T>
+  std::vector<T> read() const {
+    int len = 0;
+    ioctl(*m_socket_ptr, FIONREAD, &len);
```
`len` = amount of available data **in bytes**

tomsik68 commented on client.hpp:
```diff
@@ -13,11 +18,40 @@ class Client {
+    if (len == 0) {
+      return std::vector<T>();
+    }
+    std::vector<T> result(len);
```
receiving 3 `long`s causes `len = 3 * sizeof(long) = 24` and therefore vector is always `sizeof(T)` times longer and therefore is using unnecessary memory space.

tomsik68 reviewed (approved):
Looks good to me, checks are passing, so feel free to merge. More unit tests never hurt though ;)

tomsik68 commented on tests/tram.cpp:
```diff
@@ -55,3 +55,39 @@ TEST(tram, sendAndReceiveVector) {
+    long long numbers[] = {1, 2, 3, 4};
+    conn.write(numbers);
+    std::vector<long long> recv = client.read<long long>();
+    ASSERT_EQ(4, recv.size());
```
I think you should also check for contents of `recv`

dudoslav committed:
Fix sending data with variable length

dudoslav committed:
Add tests for long long array/vector and struct

dudoslav committed:
Rework longlongArray test to be more precise
